### PR TITLE
Feature/elineb/cvd19 1163 add parents of children

### PR DESCRIPTION
**Pull request by kathrynline**

### CODE REVIEW

File: R/parents_of_children.R [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  # Check that child_location_id is valid
+  child_hier <- hierarchy[location_id == child_location_id]  
+  if (nrow(child_hier) == 0){
+    stop("Child location is not in hierarchy!")
+  }
+  
+  # Check that child and parent are compatible
+  child_level = child_hier$level[1]
+  if (child_level <= parent_level){
+    stop(sprintf("Parent level %i and child level %i are incompatible.", parent_level, child_level))
```
Comment (epi-sam) [inline R/parents_of_children.R]: "parent level is greater than or equal to the child level" for a specific error?

File: R/parents_of_children.R [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  
+  for (parent in all_parents){
+    # Loop through all parents of this child. 
+    # If the parent you're checking exists on parent_level, then return. 
+    if (hierarchy[location_id == parent, level] == parent_level){
+      return(parent)
+    }
+  }
+  # Ideally, this never triggers, but if it does you at least have a good error message.
+  stop("Oops, something went wrong inside parents_of_children! Revisit inputs.")
```
Comment (miker985) [inline R/parents_of_children.R]: You should include the arguments in a good error message

**Review by miker985 — APPROVED** [review]

File: R/parents_of_children.R [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#' 
+#' @description Given a location ID, returns its parent ID at a given level. 
+#' 
+#' @param child_location_id [int] Location ID to pull parent ID for
+#' @param hierarchy [data.table] Hierarchy. Must have columns location_id, path_to_top_parent, and level.
+#' @param parent_level [int] Level of the parent hierarchy to query.
+#' 
+#' @export
+#' @return 
+parents_of_child <- function(
```
Comment (epi-sam) [inline R/parents_of_children.R]: prefer `parent_of_child` for clarity - each child should only return single parent at a single level

File: R/parents_of_children.R [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+#' @param child_location_id [int] Location ID to pull parent ID for
+#' @param hierarchy [data.table] Hierarchy. Must have columns location_id, path_to_top_parent, and level.
+#' @param parent_level [int] Level of the parent hierarchy to query.
+#' @export 
+parents_of_children <- function(  
+  child_loc_ids, 
+  hierarchy, 
+  parent_level
+){
+  return(sapply(child_loc_ids, function(x) parents_of_child(x, hierarchy, parent_level)))
```
Comment (epi-sam) [inline R/parents_of_children.R]: prefer `parent_of_child` for clarity

File: R/parents_of_children.R [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  }
+  # Ideally, this never triggers, but if it does you at least have a good error message.
+  stop("Oops, something went wrong inside parents_of_children! Revisit inputs.")
+}
+
+#' @description Helper function to validate inputs to function
+validate_parents_of_children_inputs <- function(child_location_id, hierarchy, parent_level){
+  # Check for valid parent_level
+  if (!parent_level %in% unique(hierarchy$level)){
```
Comment (epi-sam) [inline R/parents_of_children.R:57]: Probably also need an error if multiple `parent levels` are requested

File: R/parents_of_children.R [hunk 6]
```diff
@@ -0,0 +1,74 @@
+require(data.table)
+
+#' @name parents_of_children
+#' 
+#' @description Vectorized implementation of "parents_of_child"
+#' 
+#' @param child_location_id [int] Location ID to pull parent ID for
```
Comment (epi-sam) [inline R/parents_of_children.R]: can we match the argument name below, and clarify that a vector of location_ids can be used?

File: R/parents_of_children.R (hunk 5 above)
Comment (epi-sam) [reply R/parents_of_children.R:57]:
```
if(length(parent_level) != 1){
stop("Please specify a single parent level")
}
```

File: R/parents_of_children.R [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+#' @param hierarchy [data.table] Hierarchy. Must have columns location_id, path_to_top_parent, and level.
+#' @param parent_level [int] Level of the parent hierarchy to query.
```
Comment (epi-sam) [inline R/parents_of_children.R]: "Single level of the hierarchy - find all parent location_ids of child location_ids at this level"

File: R/parents_of_children.R (hunk 4 above)
Comment (epi-sam) [inline R/parents_of_children.R]: Should only return `unique(location_ids)`

File: R/parents_of_children.R (hunk 2 above)
Comment (kathrynline) [reply R/parents_of_children.R]: Sure; done.

File: R/parents_of_children.R (hunk 1 above)
Comment (kathrynline) [reply R/parents_of_children.R]: Thanks; done.

File: R/parents_of_children.R (hunk 3 above)
Comment (kathrynline) [reply R/parents_of_children.R]: Done.

File: R/parents_of_children.R (hunk 5 above)
Comment (kathrynline) [reply R/parents_of_children.R:57]: Thanks.

File: R/parents_of_children.R (hunk 4 above)
Comment (kathrynline) [reply R/parents_of_children.R]: Oh ok. Added a test for this.

File: R/parents_of_children.R (hunk 7 above)
Comment (kathrynline) [reply R/parents_of_children.R]: ✔️

File: R/parents_of_children.R (hunk 6 above)
Comment (kathrynline) [reply R/parents_of_children.R]: Good catch; thanks.

File: R/parents_of_children.R [hunk 8]
```diff
@@ -0,0 +1,78 @@
+#' @name parents_of_children
```
Comment (IHMEkaleb) [inline R/parents_of_children.R:1]: Looks good. Seems to cover all the bases. Thanks!

**Pull request merged**